lib/pbench/test/unit/server/query_apis/commons.py

```diff
@@ -111,43 +126,72 @@ def test_missing_json_object(self, client, server_config, pbench_token):
         assert response.status_code == HTTPStatus.BAD_REQUEST
         assert response.json.get("message") == "Invalid request payload"
 
-    def missing_keys(self, client, server_config, keys, user_ok, pbench_token):
+    def test_missing_keys(self, client, server_config, user_ok, pbench_token):
         """
         Test behavior when JSON payload does not contain all required keys.
 
         Note that Pbench will silently ignore any additional keys that are
         specified but not required.
-       """
+        """
+
+        def missing_key_helper(keys):
+            response = client.post(
+                server_config.rest_uri + self.pbench_endpoint,
+                headers={"Authorization": "Bearer " + pbench_token},
+                json=keys,
+            )
+            assert response.status_code == HTTPStatus.BAD_REQUEST
+            missing = [k for k in required_keys if k not in keys]
+            assert (
+                response.json.get("message")
+                == f"Missing required parameters: {','.join(missing)}"
+            )
+
+        parameter_items = self.cls_obj.schema.parameters.items()
+
         required_keys = [
             key
             for key, parameter in self.cls_obj.schema.parameters.items()
             if parameter.required
         ]
-        response = client.post(
-            server_config.rest_uri + self.pbench_endpoint,
-            headers={"Authorization": "Bearer " + pbench_token},
-            json=keys,
-        )
-        assert response.status_code == HTTPStatus.BAD_REQUEST
-        missing = [k for k in required_keys if k not in keys]
-        assert (
-            response.json.get("message")
-            == f"Missing required parameters: {','.join(missing)}"
-        )
+
+        all_combinations = []
+        for r in range(1, len(parameter_items) + 1):
+            combinations_object = list(itertools.combinations(parameter_items, r))
+            for item in combinations_object:
+                tmp_req_keys = []
+                for element in item:
+                    if element[1].required:
+                        tmp_req_keys.append(element[0])
+                if not sorted(tmp_req_keys) == sorted(required_keys):
+                    all_combinations.append(item)
+
+        for items in all_combinations:
+            keys = {}
+            for item in items:
+                if item[1].type == ParamType.ACCESS:
+                    keys[item[0]] = "public"
+                elif item[1].type == ParamType.DATE:
+                    keys[item[0]] = "2020"
+                else:
+                    keys[item[0]] = "foobar"
+
+            missing_key_helper(keys)
+
+        # Test for random non-existent key
+        missing_key_helper({"notakey": None})
```

Comment:
This seems misleading, since it's not the key name used here that's important but the fact that all of the required keys are missing because the arbitrary name you used is not one of them.

(And, technically, this test would be invalid if we had a query API with *no* required parameters, though I'm not sure that's likely.)

Reply:
Fixed, I added a condition that if required_keys are present then only we run this test.